src/System.ServiceModel.Primitives/tests/Channels/UnderstoodHeadersTest.cs

```diff
@@ -0,0 +1,70 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Infrastructure.Common;
+using System;
+using System.Collections;
+using System.Linq;
+using System.ServiceModel;
+using System.ServiceModel.Channels;
+using Xunit;
+
+public static class UnderstoodHeadersTest
+{
+    [WcfFact]
+    public static void UnderstoodHeaders_Add_Contains_Remove_GetEnumerator()
+    {
+        // *** SETUP *** \\
+        BasicHttpBinding binding = new BasicHttpBinding();
+
+        // Create Message object
+        Message requestMessage = Message.CreateMessage(
+            binding.MessageVersion,
+            "http://tempuri.org/AnyService",
+            new CustomBodyWriter("Any Message"));
+
+        // Create additional message headers.
+        MessageHeader customHeaderAlpha = MessageHeader.CreateHeader("AlphaHeader", "AlphaHeaderNamespace", "AlphaObjectValue");
+        MessageHeader customHeaderBravo = MessageHeader.CreateHeader("BravoHeader", "BravoHeaderNamespace", "BravoObjectValue");
+
+        // Add the new message headers to the Message
+        requestMessage.Headers.Add(customHeaderAlpha);
+        requestMessage.Headers.Add(customHeaderBravo);
+
+        // Use the "UnderstoodHeaders.Contains" API to list which MessageHeaders are set as "UnderstoodHeaders"
+        foreach (MessageHeaderInfo mhi in requestMessage.Headers)
+        {
+            // The MessageHeader with name "Action" is created with the Message object by default and should be MustUnderstand.
+            if (String.Equals(mhi.Name, "Action")) Assert.True(mhi.MustUnderstand);
+            // The two newly added headers should not have MustUnderstand set to true.
+            if (String.Equals(mhi.Name, "AlphaHeader")) Assert.True(!mhi.MustUnderstand);
+            if (String.Equals(mhi.Name, "BravoHeader")) Assert.True(!mhi.MustUnderstand);
+        }
+
+        // Use "UnderstoodHeaders.Add" API.
+        requestMessage.Headers.UnderstoodHeaders.Add(customHeaderAlpha);
+        requestMessage.Headers.UnderstoodHeaders.Add(customHeaderBravo);
+        int numUnderstoodHeaders = requestMessage.Headers.UnderstoodHeaders.Count<MessageHeaderInfo>();
+        Assert.True(numUnderstoodHeaders == 3, $"Expected 3 MessageHeaders but only found {numUnderstoodHeaders}");
+        
+        // Use "UnderstoodHeaders.GetEnumerator" API.
+        IEnumerator iEnum = requestMessage.Headers.UnderstoodHeaders.GetEnumerator();
```

Comment:
This isn't getting an IEnumerator, it's getting an IEnumerator<MessageHeaderInfo>. You should add another section to the test which casts UnderstoodHeaders to IEnumerable. E.g.:
```csharp
IEnumerator iEnum2 = ((IEnumerable)requestMessage.Headers.UnderstoodHeaders).GetEnumerator();
```